File: src/CampaignService.ts
import JSONRPCService from './JSONRPCService';

export interface AddCharacterToEncounterParams {
    EncounterID: number
    CharacterID: number
    OwnerID: number
}

export interface Empty {
}

export interface CreateCampaignParams {
    OwnerID: number
    Name: string
    Description: NullString
}

export interface Empty {
}

export interface CreateCharacterParams {
    OwnerID: number
    Name: string
    Definition: string
}

export interface Empty {
}

export interface CreateEncounterParams {
    CampaignID: number
    Name: string
    Description: NullString
    TilemapID: NullInt32
    OwnerID: number
}

export interface Empty {
}

export interface DeleteCampaignParams {
    ID: number
    OwnerID: number
}

export interface Empty {
}

export interface DeleteCharacterParams {
    ID: number
    OwnerID: number
}

export interface Empty {
}

export interface DeleteEncounterParams {
    ID: number
    OwnerID: number
}

export interface Empty {
}

export interface Empty {
}

export interface ListCampaignsResponse {
    Campaigns: FilledCampaign[] | null
}

export interface Empty {
}

export interface ListCharactersResponse {
    Characters: Character[] | null
}

export interface RemoveCharacterFromEncounterParams {
    EncounterID: number
    CharacterID: number
    OwnerID: number
}

export interface Empty {
}

export interface SearchCharactersRequest {
    Name: string
}

export interface ListCharactersResponse {
    Characters: Character[] | null
}

export interface UpdateCampaignParams {
    ID: number
    OwnerID: number
    Name: string
    Description: NullString
}

export interface Empty {
}

export interface UpdateCharacterParams {
    ID: number
    OwnerID: number
    Name: string
}

export interface Empty {
}

export interface UpdateEncounterParams {
    ID: number
    OwnerID: number
    Name: string
    Description: NullString
    TilemapID: NullInt32
}

export interface Empty {
}

export interface Campaign {
    ID: number
    OwnerID: number
    Name: string
    Description: NullString
    CreatedAt: Date
}

export interface Character {
    ID: number
    OwnerID: number
    Name: string
    Definition: string
    Sprite: string
    CreatedAt: Date
}

export interface Encounter {
    ID: number
    CampaignID: number
    Name: string
    Description: NullString
    TilemapID: NullInt32
    CreatedAt: Date
}

export interface FilledCampaign {
        
    ID: number
    OwnerID: number
    Name: string
    Description: NullString
    CreatedAt: Date

    Encounters: FilledEncounter[] | null
    Characters: Character[] | null
}

export interface FilledEncounter {
        
    ID: number
    CampaignID: number
    Name: string
    Description: NullString
    TilemapID: NullInt32
    CreatedAt: Date

    Tilemap: Tilemap
    Characters: Character[] | null
}

export interface NullInt32 {
    Int32: number
    Valid: boolean
}

export interface NullString {
    String: string
    Valid: boolean
}

export interface Tilemap {
    ID: number
    OwnerID: number
    Name: string
    Description: NullString
    Definition: string
    Hash: string
    CreatedAt: Date
}

class CampaignService extends JSONRPCService {
    async AddCharacterToEncounter(args: AddCharacterToEncounterParams): Promise<Empty> {
      return this.jsonrpc<Empty>("AddCharacterToEncounter", args);
    }
    async CreateCampaign(args: CreateCampaignParams): Promise<Empty> {
      return this.jsonrpc<Empty>("CreateCampaign", args);
    }
    async CreateCharacter(args: CreateCharacterParams): Promise<Empty> {
      return this.jsonrpc<Empty>("CreateCharacter", args);
    }
    async CreateEncounter(args: CreateEncounterParams): Promise<Empty> {
      return this.jsonrpc<Empty>("CreateEncounter", args);
    }
    async DeleteCampaign(args: DeleteCampaignParams): Promise<Empty> {
      return this.jsonrpc<Empty>("DeleteCampaign", args);
    }
    async DeleteCharacter(args: DeleteCharacterParams): Promise<Empty> {
      return this.jsonrpc<Empty>("DeleteCharacter", args);
    }
    async DeleteEncounter(args: DeleteEncounterParams): Promise<Empty> {
      return this.jsonrpc<Empty>("DeleteEncounter", args);
    }
    async ListCampaigns(args: Empty): Promise<ListCampaignsResponse> {
      return this.jsonrpc<ListCampaignsResponse>("ListCampaigns", args);
    }
    async ListCharacters(args: Empty): Promise<ListCharactersResponse> {
      return this.jsonrpc<ListCharactersResponse>("ListCharacters", args);
    }
    async RemoveCharacterFromEncounter(args: RemoveCharacterFromEncounterParams): Promise<Empty> {
      return this.jsonrpc<Empty>("RemoveCharacterFromEncounter", args);
    }
    async SearchCharacters(args: SearchCharactersRequest): Promise<ListCharactersResponse> {
      return this.jsonrpc<ListCharactersResponse>("SearchCharacters", args);
    }
    async UpdateCampaign(args: UpdateCampaignParams): Promise<Empty> {
      return this.jsonrpc<Empty>("UpdateCampaign", args);
    }
    async UpdateCharacter(args: UpdateCharacterParams): Promise<Empty> {
      return this.jsonrpc<Empty>("UpdateCharacter", args);
    }
    async UpdateEncounter(args: UpdateEncounterParams): Promise<Empty> {
      return this.jsonrpc<Empty>("UpdateEncounter", args);
    }
}

const service = new CampaignService("CampaignService");
(window as any).CampaignService = service;
export default service;

File: server/campaign_service.go
package main

import (
	"errors"
	"net/http"

	"github.com/etherealmachine/rpg.ai/server/models"
)

type CampaignService struct {
	db *models.Queries
}

type Empty struct {
}

type ListCampaignsResponse struct {
	Campaigns []models.FilledCampaign
}

func (s *CampaignService) ListCampaigns(r *http.Request, arg *Empty, reply *ListCampaignsResponse) error {
	u := currentUser(r)
	if u == nil {
		return errors.New("no authenticated user found")
	}
	reply.Campaigns = models.ListFilledCampaignsByOwnerID(r.Context(), db, u.ID)
	return nil
}

func (s *CampaignService) CreateCampaign(r *http.Request, arg *models.CreateCampaignParams, reply *Empty) error {
	u := currentUser(r)
	if u == nil {
		return errors.New("no authenticated user found")
	}
	arg.OwnerID = u.ID
	_, err := s.db.CreateCampaign(r.Context(), *arg)
	return err
}

func (s *CampaignService) DeleteCampaign(r *http.Request, arg *models.DeleteCampaignParams, reply *Empty) error {
	u := currentUser(r)
	if u == nil {
		return errors.New("no authenticated user found")
	}
	arg.OwnerID = u.ID
	return s.db.DeleteCampaign(r.Context(), *arg)
}

func (s *CampaignService) UpdateCampaign(r *http.Request, arg *models.UpdateCampaignParams, reply *Empty) error {
	u := currentUser(r)
	if u == nil {
		return errors.New("no authenticated user found")
	}
	arg.OwnerID = u.ID
	return s.db.UpdateCampaign(r.Context(), *arg)
}

type ListCharactersResponse struct {
	Characters []models.Character
}

func (s *CampaignService) ListCharacters(r *http.Request, arg *Empty, reply *ListCharactersResponse) error {
	u := currentUser(r)
	if u == nil {
		return errors.New("no authenticated user found")
	}
	var err error
	reply.Characters, err = s.db.ListCharactersByOwnerID(r.Context(), u.ID)
	return err
}

type SearchCharactersRequest struct {
	Name string
}

func (s *CampaignService) SearchCharacters(r *http.Request, arg *SearchCharactersRequest, reply *ListCharactersResponse) error {
	u := currentUser(r)
	if u == nil {
		return errors.New("no authenticated user found")
	}
	var err error
	reply.Characters, err = s.db.SearchCharacters(r.Context(), arg.Name)
	return err
}

func (s *CampaignService) CreateCharacter(r *http.Request, arg *models.CreateCharacterParams, reply *Empty) error {
	u := currentUser(r)
	if u == nil {
		return errors.New("no authenticated user found")
	}
	arg.OwnerID = u.ID
	_, err := s.db.CreateCharacter(r.Context(), *arg)
	return err
}

func (s *CampaignService) DeleteCharacter(r *http.Request, arg *models.DeleteCharacterParams, reply *Empty) error {
	u := currentUser(r)
	if u == nil {
		return errors.New("no authenticated user found")
	}
	arg.OwnerID = u.ID
	return s.db.DeleteCharacter(r.Context(), *arg)
}

func (s *CampaignService) UpdateCharacter(r *http.Request, arg *models.UpdateCharacterParams, reply *Empty) error {
	u := currentUser(r)
	if u == nil {
		return errors.New("no authenticated user found")
	}
	arg.OwnerID = u.ID
	return s.db.UpdateCharacter(r.Context(), *arg)
}

func (s *CampaignService) CreateEncounter(r *http.Request, arg *models.CreateEncounterParams, reply *Empty) error {
	u := currentUser(r)
	if u == nil {
		return errors.New("no authenticated user found")
	}
	arg.OwnerID = u.ID
	_, err := s.db.CreateEncounter(r.Context(), *arg)
	return err
}

func (s *CampaignService) DeleteEncounter(r *http.Request, arg *models.DeleteEncounterParams, reply *Empty) error {
	u := currentUser(r)
	if u == nil {
		return errors.New("no authenticated user found")
	}
	arg.OwnerID = u.ID
	return s.db.DeleteEncounter(r.Context(), *arg)
}

func (s *CampaignService) UpdateEncounter(r *http.Request, arg *models.UpdateEncounterParams, reply *Empty) error {
	u := currentUser(r)
	if u == nil {
		return errors.New("no authenticated user found")
	}
	arg.OwnerID = u.ID
	return s.db.UpdateEncounter(r.Context(), *arg)
}

func (s *CampaignService) AddCharacterToEncounter(r *http.Request, arg *models.AddCharacterToEncounterParams, reply *Empty) error {
	u := currentUser(r)
	if u == nil {
		return errors.New("no authenticated user found")
	}
	arg.OwnerID = u.ID
	return s.db.AddCharacterToEncounter(r.Context(), *arg)
}

func (s *CampaignService) RemoveCharacterFromEncounter(r *http.Request, arg *models.RemoveCharacterFromEncounterParams, reply *Empty) error {
	u := currentUser(r)
	if u == nil {
		return errors.New("no authenticated user found")
	}
	arg.OwnerID = u.ID
	return s.db.RemoveCharacterFromEncounter(r.Context(), *arg)
}
